File: backend/routers/runner.py
"""
Test Runner router — runs pytest for a given section inside a sandbox.
Returns pass/fail counts, test names, and computed score.
"""

import subprocess
import json
import os
import time
import tempfile
from fastapi import APIRouter, HTTPException, Depends
from pydantic import BaseModel
from typing import Optional

from backend.core.config import settings
from backend.core.deps import get_current_user, require_admin

router = APIRouter()

# Points per section (max)
SECTION_POINTS = {1: 100, 2: 100, 3: 100, 4: 120}

# Paths to test directories relative to repo root
SECTION_TESTS = {
    1: "section1-multifile-debug/tests",
    2: "section2-broken-recovery/tests",
    3: "section3-memory-deadlock/tests",
    4: "section4-logical-tracing/tests",
}


class RunRequest(BaseModel):
    section: int          # 1–4


class TestResult(BaseModel):
    section: int
    passed: int
    failed: int
    total: int
    score: float
    max_score: int
    duration_seconds: float
    results: list         # per-test details


@router.post("/section", response_model=TestResult)
def run_section(body: RunRequest, user: dict = Depends(require_admin)):
    """
    Run pytest for the requested section inside the repo.
    Returns pass/fail breakdown and computed score.
    """
    if body.section not in SECTION_TESTS:
        raise HTTPException(status_code=422, detail="section must be 1–4")

    test_path = os.path.normpath(
        os.path.join(settings.REPO_ROOT, SECTION_TESTS[body.section])
    )

    if not os.path.isdir(test_path):
        raise HTTPException(status_code=500, detail=f"Test directory not found: {test_path}")

    # Write JSON report to a temp file
    with tempfile.NamedTemporaryFile(suffix=".json", delete=False) as f:
        report_path = f.name

    try:
        start = time.monotonic()
        result = subprocess.run(
            [
                "python", "-m", "pytest", test_path,
                "--json-report", f"--json-report-file={report_path}",
                "--tb=no", "-q",
                f"--timeout={settings.PYTEST_TIMEOUT}",
            ],
            capture_output=True,
            text=True,
            timeout=settings.PYTEST_TIMEOUT + 10,
            cwd=settings.REPO_ROOT,
        )
        duration = time.monotonic() - start
    except subprocess.TimeoutExpired:
        raise HTTPException(status_code=408, detail="Test run timed out")
    finally:
        pass

    # Parse report
    try:
        with open(report_path) as f:
            report = json.load(f)
        os.unlink(report_path)
    except Exception:
        raise HTTPException(status_code=500, detail="Could not parse pytest report")

    summary  = report.get("summary", {})
    passed   = summary.get("passed", 0)
    failed   = summary.get("failed", 0) + summary.get("error", 0)
    total    = summary.get("total", passed + failed)

    # Score: proportional to pass rate × max section points
    max_pts  = SECTION_POINTS[body.section]
    score    = round((passed / total * max_pts) if total else 0, 2)

    # Per-test details
    def _get_test_error(t):
        for phase in ("call", "setup", "teardown"):
            pd = t.get(phase) or {}
            if pd.get("outcome") in ("failed", "error"):
                lr = pd.get("longrepr", "")
                if isinstance(lr, dict):
                    lr = lr.get("repr", "")
                return str(lr).strip()
        return ""

    tests = [
        {
            "name":     t.get("nodeid", "").split("::")[-1],
            "outcome":  t.get("outcome"),
            "duration": round(t.get("duration", 0), 3),
            "error":    _get_test_error(t) if t.get("outcome") != "passed" else "",
        }
        for t in report.get("tests", [])
    ]

    return TestResult(
        section=body.section,
        passed=passed,
        failed=failed,
        total=total,
        score=score,
        max_score=max_pts,
        duration_seconds=round(duration, 2),
        results=tests,
    )


@router.get("/sections")
def list_sections(_: dict = Depends(require_admin)):
    """Return available sections and their max points."""
    return {
        "sections": [
            {"id": s, "max_points": p, "tests_dir": SECTION_TESTS[s]}
            for s, p in SECTION_POINTS.items()
        ]
    }

File: backend/routers/runner.js
'use strict';
const router = require('express').Router();
const path   = require('path');
const fs     = require('fs');
const { spawn } = require('child_process');
const config = require('../core/config');
const { requireAdmin } = require('../middleware/auth');

const SECTION_POINTS = { 1: 100, 2: 100, 3: 100, 4: 120 };
const SECTION_TESTS  = {
  1: 'section1-multifile-debug/tests',
  2: 'section2-broken-recovery/tests',
  3: 'section3-memory-deadlock/tests',
  4: 'section4-logical-tracing/tests',
};

function runPytest(testPath, reportPath) {
  return new Promise((resolve, reject) => {
    const timeout = (config.PYTEST_TIMEOUT + 10) * 1000;
    const proc = spawn('python3', [
      '-m', 'pytest', testPath,
      '--json-report', `--json-report-file=${reportPath}`,
      '--tb=no', '-q',
      `--timeout=${config.PYTEST_TIMEOUT}`,
    ], { cwd: config.REPO_ROOT });

    let killed = false;
    const timer = setTimeout(() => {
      killed = true;
      proc.kill('SIGKILL');
      reject(new Error('timeout'));
    }, timeout);

    proc.on('close', () => {
      clearTimeout(timer);
      if (!killed) resolve();
    });
    proc.on('error', err => { clearTimeout(timer); reject(err); });
  });
}

// POST /api/run/section
router.post('/section', requireAdmin, async (req, res, next) => {
  try {
    const section = parseInt(req.body.section, 10);
    if (!SECTION_TESTS[section])
      return res.status(422).json({ detail: 'section must be 1–4' });

    const testPath   = path.normalize(path.join(config.REPO_ROOT, SECTION_TESTS[section]));
    if (!fs.existsSync(testPath))
      return res.status(500).json({ detail: `Test directory not found: ${testPath}` });

    const reportPath = path.join(require('os').tmpdir(), `dc_run_${Date.now()}.json`);
    const startTs    = Date.now();

    try {
      await runPytest(testPath, reportPath);
    } catch (err) {
      if (err.message === 'timeout') return res.status(408).json({ detail: 'Test run timed out' });
      return next(err);
    }
    const duration = (Date.now() - startTs) / 1000;

    let report;
    try {
      report = JSON.parse(fs.readFileSync(reportPath, 'utf8'));
      fs.unlinkSync(reportPath);
    } catch {
      return res.status(500).json({ detail: 'Could not parse pytest report' });
    }

    const summary  = report.summary || {};
    const passed   = summary.passed || 0;
    const failed   = (summary.failed || 0) + (summary.error || 0);
    const total    = summary.total  || passed + failed;
    const maxPts   = SECTION_POINTS[section];
    const score    = total ? Math.round((passed / total) * maxPts * 100) / 100 : 0;

    const tests = (report.tests || []).map(t => {
      let error = '';
      for (const phase of ['call', 'setup', 'teardown']) {
        const pd = t[phase] || {};
        if (['failed', 'error'].includes(pd.outcome)) {
          error = typeof pd.longrepr === 'object' ? (pd.longrepr.repr || '') : (pd.longrepr || '');
          break;
        }
      }
      return {
        name:     (t.nodeid || '').split('::').pop(),
        outcome:  t.outcome,
        duration: Math.round((t.duration || 0) * 1000) / 1000,
        error:    t.outcome !== 'passed' ? error.trim() : '',
      };
    });

    res.json({ section, passed, failed, total, score, max_score: maxPts,
               duration_seconds: Math.round(duration * 100) / 100, results: tests });
  } catch (err) { next(err); }
});

// GET /api/run/sections
router.get('/sections', requireAdmin, (_req, res) => {
  res.json({
    sections: Object.entries(SECTION_POINTS).map(([id, pts]) => ({
      id: parseInt(id, 10), max_points: pts, tests_dir: SECTION_TESTS[parseInt(id, 10)],
    })),
  });
});

module.exports = router;
